miden/src/repl/mod.rs

```diff
@@ -0,0 +1,317 @@
+                } else if line.len() > 6 && &line[..5] == "!mem[" {
+                    // if user wants to see the state of a particular address in a memory, the input should be atleast
+                    // of length 5 and first four characters should be "!mem[" and should end with "]".
+
+                    // extracting the digits from the input text given by the user.
+                    let (index, remainder) = extract_digits(&line[5..]);
+
+                    // convert the parsed digits in integer form.
+                    let memory_at_addr: u64 = index.parse().unwrap();
+
+                    let mut flag = false;
+
+                    if remainder == "]" {
+                        for (i, memory_value) in &memory {
+                            if *i == memory_at_addr {
+                                // each memory register converted from Felt to an integer.
+                                let mem_int =
+                                    memory_value.iter().map(|&x| x.as_int()).collect::<Vec<_>>();
+                                // displays the state of the memory at index x if present.
+                                println!("{}: {:?}", memory_at_addr, mem_int);
+                                flag = true;
+                                break;
+                            }
+                        }
+                        if !flag {
+                            println!("Memory at address {} is empty", memory_at_addr,);
+                        }
+                    }
+                    stack_flag = false;
+                } else if line == "!undo" {
+                    let last_line = program_lines.pop().unwrap();
+                    println!("Undoing {}", last_line);
+                    stack_flag = true;
+                } else if line == "!stack" {
+                    println!("{}", result_string);
+                    stack_flag = false;
+                } else {
+                    rl.add_history_entry(line.clone());
+                    program_lines.push(line.clone());
+                    stack_flag = true;
+                }
+            }
+            Err(ReadlineError::Interrupted) => {
+                println!("CTRL-C");
+                break;
+            }
+            Err(ReadlineError::Eof) => {
+                println!("CTRL-D");
+                break;
+            }
+            Err(err) => {
+                println!("Error: {:?}", err);
+                break;
+            }
+        };
+    }
+    rl.save_history("history.txt").unwrap();
+}
+
+/// HELPER METHODS
+/// --------------------------------------------------------------------------------------------
+
+/// Compiles and executes a compiled Miden program, returning the stack, memory and any Miden errors.
+/// The program is passed in as a String, passed to the Miden Assembler, and then passed into the Miden
+/// Processor to be executed.
+pub fn execute(
+    program: String,
+    _pub_inputs: Vec<u128>,
```

Comment:
Does this function need to be public?

Also, seems like `pub_inputs` are unused, do we need this param?